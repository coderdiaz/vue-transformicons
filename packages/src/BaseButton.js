export const BaseButton = {
  name: 'BaseButton',
  props: {
    animation: {
      type: String,
      required: true,
    },
    ariaLabel: {
      type: String,
      required: true,
    },
    active: {
      type: Boolean,
      default: false,
    },
    animations: {
      type: Object,
      default: {},
    },
  },
  data() {
    return {
      currentActive: false,
    };
  },
  mounted() {
    this.currentActive = this.active;
  },
  computed: {
    currentClasses() {
      const currentAnimation = this.animations[this.animation].join(' ');
      return {
        tcon: true,
        [`${currentAnimation}`]: true,
        'tcon-transform': (this.currentActive),
      };
    },
  },
  render() {
    return <button
      aria-label={this.ariaLabel}
      class={this.currentClasses}
      onclick={this.toggleStatus}>
      {this.$slots.default}
    </button>;
  },
  methods: {
    toggleStatus(e) {
      e.preventDefault();
      this.currentActive = !this.currentActive;
    },
  },
};

export const IconMixin = {
  props: {
    animation: {
      type: String,
      default: '',
    },
    active: {
      type: Boolean,
      default: false,
    },
  },
};
