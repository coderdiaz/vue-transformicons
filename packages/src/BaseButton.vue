<script>
export default {
  name: 'BaseButton',
  props: {
    animation: {
      type: String,
      required: true,
    },
    ariaLabel: {
      type: String,
      required: true,
    },
    active: {
      type: Boolean,
      default: false,
    },
    animations: {
      type: Object,
      required: true,
      validator(value) {
        return Object.keys(value).length !== 0;
      },
    },
  },
  data() {
    return {
      currentActive: false,
    };
  },
  mounted() {
    this.currentActive = this.active;
  },
  computed: {
    currentClasses() {
      const currentAnimation = this.animations[this.animation].join(' ');
      return {
        tcon: true,
        [`${currentAnimation}`]: true,
        'tcon-transform': (this.currentActive),
      };
    },
  },
  render() {
    return <button
      aria-label={this.ariaLabel}
      class={this.currentClasses}
      onClick={this.toggleStatus}>
      {this.$slots.default}
    </button>;
  },
  methods: {
    toggleStatus(e) {
      e.preventDefault();
      this.currentActive = !this.currentActive;
    },
  },
  watch: {
    active: {
      handler(val) {
        this.currentActive = val;
      },
    },
  },
};
</script>
<style src="../transformicons.css">
