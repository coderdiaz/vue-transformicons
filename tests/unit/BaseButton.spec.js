import { shallowMount } from '@vue/test-utils';
import BaseButton from '../../packages/src/BaseButton.vue';

const vm = shallowMount(BaseButton, {
  propsData: {
    animation: 'minus',
    ariaLabel: 'add item',
    animations: {
      minus: ['tcon-plus', 'tcon-plus--minus'],
    },
  },
});

// Helper function to create a component
const createComponent = (propsData, slots) => shallowMount(BaseButton, { propsData, slots });

describe('BaseButton.vue', () => {
  let cmp;

  it('should match the snapshot', () => {
    expect(vm.$el).toMatchSnapshot();
  });

  it('has a mounted hook', () => {
    expect(typeof BaseButton.mounted).toBe('function');
  });

  describe('Slots', () => {
    it('should insert the fake element in the default slot', () => {
      const component = createComponent({
        animation: 'minus',
        ariaLabel: 'add item',
        animations: {
          minus: ['tcon-plus', 'tcon-plus--minus'],
        },
      }, {
        default: '<div class="fake-element"></div>',
      });
      expect(component.findAll('.fake-element').length).toBe(1);
    });
  });

  describe('Properties', () => {
    it('should set the active prop when the component is mounted', () => {
      const component = createComponent({
        animation: 'minus',
        ariaLabel: 'add item',
        active: true,
        animations: {
          minus: ['tcon-plus', 'tcon-plus--minus'],
        },
      });

      expect(component.vm.currentActive).toBeTruthy();
    });
  });

  describe('Methods', () => {
    beforeEach(() => {
      cmp = createComponent({
        animation: 'minus',
        ariaLabel: 'add item',
        active: false,
        animations: {
          minus: ['tcon-plus', 'tcon-plus--minus'],
        },
      });
    });

    it('should change the current active status when toggleStatus method is triggered', () => {
      expect(cmp.vm.currentActive).toBeFalsy();
      cmp.find('button').trigger('click');
      expect(cmp.vm.currentActive).toBeTruthy();
    });
  });
});
