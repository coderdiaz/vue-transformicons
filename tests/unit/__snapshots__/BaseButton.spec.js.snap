// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseButton.vue should match the snapshot 1`] = `undefined`;
